File: 2020/day3.py
data = ""
with open("inputs/day3.txt", "r") as f:
    data = f.read()

rows = data.split("\n")


def Traverse(y: int, x: int):
    index = 0
    trees = 0
    for row in rows[::y]:
        index %= len(row)
        if row[index] == "#":
            trees += 1
        index += x
    return trees


def Part1() -> None:
    print("Trees Encountered: {}".format(Traverse(1, 3)))


def Part2() -> None:
    print(
        "Multiplied Trees: {}".format(
            Traverse(1, 1)
            * Traverse(1, 3)
            * Traverse(1, 5)
            * Traverse(1, 7)
            * Traverse(2, 1)
        )
    )


def main() -> None:
    print("Day 3")
    Part1()
    Part2()


if __name__ == "__main__":
    main()

File: 2015/day3.ts
import { readFileSync } from "fs";

const data = readFileSync("./inputs/day3.txt", "utf-8");

class Point {
  constructor(readonly x: number, readonly y: number) {}

  getNext(val: string) {
    switch (val) {
      case "^":
        return new Point(this.x, this.y + 1);
      case "v":
        return new Point(this.x, this.y - 1);
      case ">":
        return new Point(this.x + 1, this.y);
      case "<":
        return new Point(this.x - 1, this.y);
      default:
        throw Error(`Unknown direction: ${val}`);
    }
  }

  toString() {
    return `[${this.x},${this.y}]`;
  }
}

function part1() {
  const houseDeliveries: Map<string, number> = data.split("").reduce(
    (acc, val) => {
      acc.last = acc.last.getNext(val);
      const key = acc.last.toString();
      if (!acc.visits.has(key)) {
        acc.visits.set(key, 0);
      }
      acc.visits.set(key, acc.visits.get(key)! + 1);
      return acc;
    },
    {
      visits: new Map([["[0,0]", 1]]),
      last: new Point(0, 0),
    }
  ).visits;
  console.log(`Houses receiving presents: ${houseDeliveries.size}`);
}

function part2() {
  const houseDeliveries: Map<string, number> = data.split("").reduce(
    (acc, val) => {
      let key = "";
      if (acc.santaTurn) {
        acc.santaLast = acc.santaLast.getNext(val);
        key = acc.santaLast.toString();
      } else {
        acc.roboLast = acc.roboLast.getNext(val);
        key = acc.roboLast.toString();
      }
      acc.santaTurn = !acc.santaTurn;
      if (!acc.visits.has(key)) {
        acc.visits.set(key, 0);
      }
      acc.visits.set(key, acc.visits.get(key)! + 1);
      return acc;
    },
    {
      visits: new Map([["[0,0]", 2]]),
      santaLast: new Point(0, 0),
      roboLast: new Point(0, 0),
      santaTurn: true,
    }
  ).visits;
  console.log(`Houses receiving presents: ${houseDeliveries.size}`);
}

console.log("DAY 3");
part1();
part2();
